securitas/controller/registration.py

```diff
@@ -1,48 +1,63 @@
-from flask import flash, redirect, request, url_for
+from flask import flash, redirect, url_for, render_template
 import python_freeipa
 
 from securitas import app, ipa_admin
+from securitas.form.register_user import RegisterUserForm
 from securitas.security.ipa import untouched_ipa_client
 
 
-@app.route('/register', methods=['POST'])
+@app.route('/register', methods=['GET', 'POST'])
 def register():
-    first_name = request.form.get('first_name')
-    last_name = request.form.get('last_name')
-    username = request.form.get('username')
-    password = request.form.get('password')
-    password_confirm = request.form.get('password_confirm')
-
-    if not all([first_name, last_name, username, password, password_confirm]):
-        flash('Please fill in all fields to register an account.', 'red')
-        return redirect(url_for('root'))
-
-    if password != password_confirm:
-        flash('Password and confirmation did not match.', 'red')
-        return redirect(url_for('root'))
-
-    try:
-        ipa_admin.user_add(
-            username,
-            first_name,
-            last_name,
-            '%s %s' % (first_name, last_name),  # TODO ???
-            user_password=password,
-            login_shell='/bin/bash',
-        )
-
-        # Now we fake a password change, so that it's not immediately expired.
-        # This also logs the user in right away.
-        ipa = untouched_ipa_client(app)
-        ipa.change_password(username, password, password)
-    except python_freeipa.exceptions.FreeIPAError as e:
-        print(e)
-        flash('An error occurred while creating the account, please try again.', 'red')
-        return redirect(url_for('root'))
-
-    flash(
-        'Congratulations, you now have an account! Go ahead and sign in to proceed.',
-        'green',
-    )
-
-    return redirect(url_for('root'))
+    form = RegisterUserForm()
+
+    if form.validate_on_submit():
+        username = form.username.data
+        password = form.password.data
+        try:
+            ipa_admin.user_add(
+                username,
+                form.firstname.data,
+                form.lastname.data,
+                f'{form.firstname.data} {form.lastname.data}',  # TODO ???
+                user_password=password,
+                login_shell='/bin/bash',
+            )
+
+            # Now we fake a password change, so that it's not immediately expired.
+            # This also logs the user in right away.
+            ipa = untouched_ipa_client(app)
+            ipa.change_password(username, password, password)
+        except python_freeipa.exceptions.PWChangePolicyError as e:
+            # situations like "password is too short"
+            form.password.errors.append(e.policy_error)
+            ipa_admin.user_del(username)
```

Comment:
@abompard

Sorry, but what problem is this (and the one below) solving? What happens if I give a username that already exists, but also a short password, are you 150% sure that `DuplicateEntry` gets thrown before `PWChangePolicyError` or is there a risk that the uer account gets deleted? Are you 150% sure (below) that I can't trigger a `FreeIPAError` before a `DuplicatEntry` based one some bad field entry I make?

I think we should *never* `user_del` *anywhere* in the codebase, and I think it should be VERY rare that we use `ipa_admin` (which, in production, will not have permission to `user_del` anyway, it will be locked down to have only the ability to change passwords and create accounts).

Reply:
> Sorry, but what problem is this (and the one below) solving?

Before this change, when a user registers with a password that does not follow the policy, they get an error message about being unable to register. But the user is actually created, it's just the password change that failed, so it's pretty confusing. 

As for the one below (I suppose you mean catching the generic `FreeIPAError`), it's about catching other errors that happen during the password change process and not leaving the user in a similar state.

> What happens if I give a username that already exists, but also a short password, are you 150% sure that `DuplicateEntry` gets thrown before `PWChangePolicyError` in _every_ possible case or is there a risk that the user account gets deleted?

Fair question, the `DuplicateEntry` exception will be raised by the `user_add` method, and only `change_password` raises a `PWChangePolicyError`, so you can't reach that error handler on an existing account.

> Are you 150% sure (below) that I can't trigger a `FreeIPAError` before a `DuplicatEntry` based one some bad field entry I make?

Ah, I did not think of that one, and `FreeIPAError` is indeed a pretty generic error so I think it could happen (and if not now it may happen in future evolutions of the library). I agree that this one should go.
It's probably a bad idea to have common error handlers for both method calls anyway.

> I think we should _never_ `user_del` _anywhere_ in the codebase, and I think it should be VERY rare that we use `ipa_admin` (which, in production, will not have permission to `user_del` anyway, it will be locked down to have only the ability to change passwords and create accounts).

That's an interesting piece of information indeed :-) Then we must find a way to not leave users out in the cold when they pick a password that does not comply with the policy. I'll think about it and send a PR.

Thanks for the comments.

Reply:
OK I've sent #93, please tell me what you think.